### DISCUSSION TITLE
State without clone

### DISCUSSION THREAD

**Discussion by Moggers** [tokio-rs/axum#2821] [Q&A] [answered]
### Summary

[The docs for State](https://docs.rs/axum/latest/axum/extract/struct.State.html#when-states-need-to-implement-clone
) specify that a top level state *may* be `!Clone` if handlers never try to extract the state itself, and only use the inner values of the state by implementing `FromRef<OuterState> for InnerState`
However [with_state](https://docs.rs/axum/latest/axum/routing/struct.Router.html#method.with_state) takes an `S: Clone`.
So how do you actually insert a state which is `!Clone` into a router when you intend to only extract the inner states?

### axum version

0.7.5

**Comment by jplatte** [comment, marked as answer]
This is a documentation bug. We used to store the state inside of an `Arc` implicitly, but we no longer do so that section of the docs is just wrong. Do you want to send a PR to remove that section?

**Comment by Moggers** [comment]
Ah, ta.
Will put something up. Probably tomorrow though. 3:30am where I am right now :zzz:

**Reply by jplatte** [reply]
I think this hasn't happened, right? No obligations of course, just wanted to remind.

**Reply by yanns** [reply]
The [documentation](https://docs.rs/axum/latest/axum/extract/struct.State.html#when-states-need-to-implement-clone) now states:
> Your top level state type must implement Clone to be extractable with State:

**Reply by jplatte** [reply]
It also (still) states that

> This is also true if you’re extracting substates, _unless_ you never extract the top level state itself:

This discussion is about that "unless" part, which hasn't been true for quite a while.

**Reply by yanns** [reply]
ah ok, I missed that. So you should remove this part, and just document that the state should implement `Clone`.

**Reply by jplatte** [reply]
Yup. Guess I'll make a PR.

**Reply by jplatte** [reply]
Tomorrow.. 😅

**Reply by yanns** [reply]
Sorry, I've used "you" but I wanted to write "we"... No pressure for a PR. I can also take care of it.

**Reply by yanns** [reply]
https://github.com/tokio-rs/axum/pull/2972
